### PR TITLE
cincinnati/plugins/edge_add_remove: change, document, and test label precedence

### PR DESCRIPTION
**Pull request by steveej**
/cc @shiywang @lucab 

FYI @wking

### CODE REVIEW

File: cincinnati/src/plugins/internal/edge_add_remove.rs [hunk 1]
```diff
@@ -423,4 +434,81 @@ mod tests {
...
+            (
+                1,
+                [
+                    (
+                        format!("{}.{}", key_prefix, "previous.remove"),
+                        String::from("0.0.0"),
+                    ),
+                    (
+                        format!("{}.{}", key_prefix, "next.add"),
+                        String::from("2.0.0"),
```
Comment (lucab) [inline cincinnati/src/plugins/internal/edge_add_remove.rs]: I may be misunderstanding something. Shouldn't this be supposed to cover (i.e. re-add) the same edge that the directive above removes?

**Review by lucab — COMMENTED** [review]
Thanks, very valuable to have in an explicit form! I just have a doubt on what's the effective coverage of the new test.

File: cincinnati/src/plugins/internal/edge_add_remove.rs (hunk 1 above)
Comment (steveej) [reply cincinnati/src/plugins/internal/edge_add_remove.rs]: This label is contradicted by the subsequent label at https://github.com/openshift/cincinnati/pull/127/files#diff-0b0028d0d2d6c8d8c41ae291b2dce0ebR474.

I could add more contradictions to an extent where every label has one. Would you prefer that?

Comment (lucab) [reply cincinnati/src/plugins/internal/edge_add_remove.rs]: I see.
My understanding was that this was only covering contradicting metadata processing **for each node**, while the case seems to relies on contradicting metadata across **two different nodes**. Shallow reading from my side, sorry.

This test is fine then, maybe an hint in the docstring would make the above obvious.

File: cincinnati/src/plugins/internal/edge_add_remove.rs [hunk 2]
```diff
@@ -41,6 +41,17 @@ impl PluginSettings for EdgeAddRemovePlugin {
 /// Adds and removes next and previous releases specified by metadata.
 ///
 /// The labels are assumed to have the syntax `<prefix>.(previous|next).(remove|add)=(<Version>,)*<Version>`
+///
+/// # Label processing order
+/// The labels are processed in the following order:
```
Comment (lucab) [inline cincinnati/src/plugins/internal/edge_add_remove.rs]: Maybe `Labels are processed in multiple graph-wide passes, in the following order:`?

File: cincinnati/src/plugins/internal/edge_add_remove.rs (hunk 1 above)
Comment (steveej) [reply cincinnati/src/plugins/internal/edge_add_remove.rs]: That is an interesting twist which I didn't intend to cover in this test. I will go back and think about incorporating that one too. I'll add a note to the docstring either way.

**Comment by steveej** [comment]
@lucab hopefully all comments are addressed to both of our understand now. PTAL

File: cincinnati/src/plugins/internal/edge_add_remove.rs [hunk 3]
```diff
@@ -41,6 +41,14 @@ impl PluginSettings for EdgeAddRemovePlugin {
 /// Adds and removes next and previous releases specified by metadata.
 ///
 /// The labels are assumed to have the syntax `<prefix>.(previous|next).(remove|add)=(<Version>,)*<Version>`
+///
+/// # Label processing order
+/// The labels are grouped and processed in two separate passes in the following:
```
Comment (lucab) [inline cincinnati/src/plugins/internal/edge_add_remove.rs]: "in the following order"

**Comment by lucab** [comment]
Small typo, all fine otherwise. I'll leave this pending so that @wking has time to catch up and raise concerns (if any) on the semantics break.

/hold
/lgtm

**Comment by lucab** [comment]
/lgtm

(Hold still in place on purpose)

**Comment by steveej** [comment]
Thanks for the review @lucab. 

I'm fine merging this without further discussion, because the original ordering was a random choice at the time of implementing it, and the code change actually comes from a place of thought. I observed no significant\* difference in the result compared to the production instance so consider this change safe.

/hold cancel

\* the output is not bit-equal due to random re-ordering of JSON dictionary entries.

**Pull request merged**